compiler/grainlsp/utils.re

```diff
@@ -0,0 +1,816 @@
+        get_node_from_expression(~line, ~char, block);
+      } else {
+        Expression(expr);
+      }
+
+    | TExpFor(e1, e2, e3, block) =>
+      if (is_point_inside_location(~line, ~char, block.exp_loc)) {
+        get_node_from_expression(~line, ~char, block);
+      } else {
+        List.fold_left(
+          (acc, exp) =>
+            switch (exp) {
+            | None => acc
+            | Some(e: Typedtree.expression) =>
+              if (is_point_inside_location(~line, ~char, e.exp_loc)) {
+                get_node_from_expression(~line, ~char, e);
+              } else {
+                acc;
+              }
+            },
+          Expression(expr),
+          [e1, e2, e3],
+        );
+      }
+
+    | TExpConstruct(_, _, expressions) =>
+      find_location_in_expressions(
+        ~line,
+        ~char,
+        ~default=Expression(expr),
+        expressions,
+      )
+    };
+
+  switch (node) {
+  | NotInRange
+  | Error(_) =>
+    if (is_point_inside_location(~line, ~char, expr.exp_loc)) {
+      Expression(expr);
+    } else {
+      NotInRange;
+    }
+  | _ => node
+  };
+};
+
+let get_node_from_statement =
+    (~line, ~char, stmt: Grain_typed__Typedtree.toplevel_stmt) => {
+  switch (stmt.ttop_desc) {
+  | TTopImport(import_declaration) => Error("import declaration")
+  | TTopForeign(export_flag, value_description) => Error("foreign")
+  | TTopData(data_declarations) => Error("data declaration")
+  | TTopLet(export_flag, rec_flag, mut_flag, value_bindings) =>
+    switch (value_bindings) {
+    | [] => Error("")
+    | _ =>
+      let matches =
+        List.map(
+          (vb: Typedtree.value_binding) =>
+            get_node_from_expression(~line, ~char, vb.vb_expr),
+          value_bindings,
+        );
+      let filtered =
+        List.filter(
+          m =>
+            switch (m) {
+            | Error(_) => false
+            | _ => true
+            },
+          matches,
+        );
+      switch (filtered) {
+      | [] =>
+        // return the type for the whole statement
+        let vb = List.hd(value_bindings);
+        let expr = vb.vb_expr;
+        Expression(expr);
+      | [node] => node
+      | _ => Error("Too many ttoplet matches")
+      };
+    }
+
+  | TTopExpr(expression) =>
+    get_node_from_expression(~line, ~char, expression)
+  | TTopException(export_flag, type_exception) => Error("exception")
+  | TTopExport(export_declarations) => Error("export")
+  };
+};
+
+let find_completable = (text, offset) => {
+  let rec loop = i => {
+    i < 0
+      ? Lident(String.sub(text, i + 1, offset - (i + 1)))
+      : (
+        switch (text.[i]) {
+        | 'a' .. 'z'
+        | 'A' .. 'Z'
+        | '0' .. '9'
+        | '.'
+        | '_' => loop(i - 1)
+        | _ =>
+          i == offset - 1
+            ? Nothing : Lident(String.sub(text, i + 1, offset - (i + 1)))
+        }
+      );
+  };
+  loop(offset - 1);
```

Comment:
This function hurts my head

Reply:
Mine too, it's taken from the Reason LSP